File: RavelinCoreDemoPodsObjC/RavelinCoreDemoPodsObjC/ViewController.m
//
//  ViewController.m
//  RavelinCoreDemoPodsObjC

#import "ViewController.h"

@import RavelinCore;
@interface ViewController ()
@property (strong, nonatomic) Ravelin *ravelin;
@property (weak, nonatomic) IBOutlet UITextField *textField;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = UIColor.lightGrayColor;
    self.textField.delegate = self;
}

- (IBAction)touchTrackButton:(id)sender {
    [self useCore];
}

- (void)useCore {

    // Make Ravelin instance with API keys
    // to assist with early stage development/debug
    // execute 'python3 simple_server.py' in a terminal session
    // and use "local" as the apiKey
    // for further development and release, replace with your Ravelin Publishable API Key
    // self.ravelin = [Ravelin createInstance:@"publishable_key_live_----"];
    self.ravelin = [Ravelin createInstance:@"local"];

    self.ravelin.customerId = @"customer012345";

    // Send a device fingerprint with a completion block
    [self.ravelin trackFingerprint:^(NSData *data, NSURLResponse *response, NSError *error) {
        if(!error) {
            NSDictionary *responseData;
            NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
            if (httpResponse.statusCode == 200) {
                responseData = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers|NSJSONReadingAllowFragments error:nil];
                NSLog(@"trackFingerprint - success");
                // track login
                [self.ravelin trackLogin:@"login"];
                self.ravelin.orderId = @"order-001";
                // track customer moving to a new page
                [self.ravelin trackPage:@"checkout"];
                // track logout
                [self.ravelin trackLogout:@"logout"];
            } else {
                // Status was not 200. Handle failure
                NSLog(@"trackFingerprint - failure");
            }
        } else {
            NSLog(@"%@",error.localizedDescription);
        }
    }];
}

- (BOOL)textField:(UITextField *)iTextField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
{
    // avoid checking the pasteboard when a user is typing and only adding a single character at a time
    if(string.length <= 1) { return YES; }
    // Check if the textfield contains pasted text
    if([string containsString:[UIPasteboard generalPasteboard].string]) {
        [self.ravelin trackPaste:@"home" pastedValue:string];
    }

    return YES;
}

@end

File: Shared/ViewController.swift
//
//  ViewController.swift
//  RavelinCoreDemo

import UIKit

// to assist with early stage development/debug
// execute 'python3 simple_server.py' in a terminal session
// and use "local" as the apiKey
let apiKey = "local"
// for further development and release, replace with your Ravelin Publishable API Key
// let apiKey = "publishable_key_live_----"

class ViewController: UIViewController {
    
    private var sdkHelper: SDKHelper?
    private var activityIndicatorViewController: ActivityIndicatorViewController?

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .lightGray
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        setupSDKHelper()
    }
    
    func setupSDKHelper() {
        guard sdkHelper == nil else {
            logInfo("sdkHelper already initialised")
            return
        }
        addActivityIndicator()
        sdkHelper = SDKHelper(apiKey: apiKey, customerId: "cust-01234", appVersion: Bundle.main.version) {_ in
            self.removeActivityIndicator()
        }
    }
    
    func trackLogin() {
        guard let sdkHelper else {
            logInfo("sdkHelper not initialised")
            return
        }
        addActivityIndicator()
        sdkHelper.trackLogin {_ in
            self.removeActivityIndicator()
        }
    }
    
    @IBAction func loginButtonTapped(_ sender: Any) {
        trackLogin()
    }
    
    func addActivityIndicator() {
        addChildViewController(ActivityIndicatorViewController()) { viewController in
            self.activityIndicatorViewController = viewController as? ActivityIndicatorViewController
        }
    }
    
    func removeActivityIndicator(after: Double = 0.5) {
        removeChildViewController(activityIndicatorViewController, after: after) {
            self.activityIndicatorViewController = nil
        }
    }
}
